bin/population_template

```diff
@@ -84,168 +89,613 @@ except NameError:
   pass
 
 
-
 def abspath(*arg):
   return os.path.abspath(os.path.join(*arg))
 
+
 def relpath(*arg):
   from mrtrix3 import app #pylint: disable=no-name-in-module, import-outside-toplevel
   return os.path.relpath(os.path.join(*arg), app.WORKING_DIR)
 
-def check_linear_transformation (transformation, cmd, max_scaling = 0.5, max_shear = 0.2, max_rot = None, pause_on_warn = True):
-  from mrtrix3 import app, run, utils #pylint: disable=no-name-in-module, import-outside-toplevel
+def copy(src, dst, follow_symlinks=True):
+  """Copy data but do not set mode bits. Return the file's destination.
 
+  mimics shutil.copy but without setting mode bits as shutil.copymode can fail on exotic mounts
+  (observed on cifs with file_mode=0777).
+  """
+  if os.path.isdir(dst):
+    dst = os.path.join(dst, os.path.basename(src))
+  if sys.version_info[0] > 2:
+    shutil.copyfile(src, dst, follow_symlinks=follow_symlinks)   # pylint: disable=unexpected-keyword-arg
+  else:
+    shutil.copyfile(src, dst)
+  return dst
+
+
+def check_linear_transformation(transformation, cmd, max_scaling=0.5, max_shear=0.2, max_rot=None, pause_on_warn=True):
+  from mrtrix3 import app, run #pylint: disable=no-name-in-module, import-outside-toplevel
   if max_rot is None:
     max_rot = 2 * math.pi
+  def load_key_value(file_path):
+    res = {}
+    with open(file_path, "r") as ftrafo:
+      for line in ftrafo.readlines():
+        if len(line) == 1 or line.startswith("#"):
+          continue
+        name, var = line.rstrip().partition(":")[::2]
+        if name in res.keys():
+          res[name].append(var.split())
+        else:
+          res[name] = var.split()
+    return res
 
   good = True
   run.command('transformcalc ' + transformation + ' decompose ' + transformation + 'decomp')
-  if not os.path.isfile(transformation + 'decomp'): # does not exist if run with -continue option
+  if not os.path.isfile(transformation + 'decomp'):  # does not exist if run with -continue option
     app.console(transformation + 'decomp not found. skipping check')
     return True
-  data = utils.load_keyval(transformation + 'decomp')
```

Comment:
What's the rationale for `load_key_value()` rather than `utils.load_keyval()`? It seems to be doing an almost identical function, except that the version here is skipping lines commencing with "`#`" (which is where key-values go in text files). If there's some unmet need, would rather address it in the library function.

Reply:
The reason that this does not use `utils.load_keyval` is that this code is 3 years older (0f85864c542908e7330c0e1e). It's still there because the output format of `transformcalc decompose` is not the key-value format. We might want to change that?
The current format is:
```text
# transformcalc trafo.txt decompose trafo.dec  (version=3.0_RC3-1182-gcbcb065d)
scaling: 1 0.999999999999999 0.999999999999999
shear: 3.05311331771918e-16 -8.32667268468867e-17 1.66533453693773e-16
angles: 0.443338396922925 0.0206736092750134 -0.0770429056771817
angle_axis: 0.449556495888172 0.983870564432754 0.0831931410457268 -0.158359128962367
translation: 0.114196353312504 -1.08436236954077 -0.922431985054211
R: 0.996820605461134 0.0769502650670704 0.0206721366628588
R: -0.0606847655421982 0.901327603695001 -0.428865841550147
R: -0.051633707587086 0.426247824068974 0.903131636428088
S: 1 3.05311331771918e-16 -8.32667268468867e-17
S: 1.87350135405495e-16 0.999999999999999 1.66533453693773e-16
S: -2.70616862252382e-16 1.11022302462516e-16 0.999999999999999
jacobian_det: 1
```

Reply:
Never mind, I misread the code of `utils.load_keyval`. Change is in a21cebd07788

Reply:
Actually that was me forgetting about the nature of the output of `transformcalc decompose`: `utils.load_keyval()` is supposed to only load key-value pairs from commented lines. Maybe if you were to provide kwarg "`comments=''`", and then explicitly remove any commented lines yourself, it might work? Or the internals of `utils.load_keyval()` could be tweaked to make this work?

I don't think the output of `transformcalc decompose` should be changed, since those key-value pairs are the principal output of the command rather than sidecar information, so it makes sense to not be commented.

Reply:
Seems to be a feature / bug in `utils.load_keyval` :P
```
utils.load_keyval('trafo.dec')
Out[2]:
{'scaling': ['1', '0.999999999999999', '0.999999999999999'],
 'shear': ['3.05311331771918e-16',
  '-8.32667268468867e-17',
  '1.66533453693773e-16'],
 'angles': ['0.443338396922925', '0.0206736092750134', '-0.0770429056771817'],
 'angle_axis': ['0.449556495888172',
  '0.983870564432754',
  '0.0831931410457268',
  '-0.158359128962367'],
 'translation': ['0.114196353312504',
  '-1.08436236954077',
  '-0.922431985054211'],
 'R': ['0.996820605461134',
  '0.0769502650670704',
  '0.0206721366628588',
  ['-0.0606847655421982', '0.901327603695001', '-0.428865841550147'],
  ['-0.051633707587086', '0.426247824068974', '0.903131636428088']],
 'S': ['1',
  '3.05311331771918e-16',
  '-8.32667268468867e-17',
  ['1.87350135405495e-16', '0.999999999999999', '1.66533453693773e-16'],
  ['-2.70616862252382e-16', '1.11022302462516e-16', '0.999999999999999']],
 'jacobian_det': ['1']}
```

Reply:
It's probably something that I didn't think through enough, since this is the only location where it's used. I probably also wrote it trying to use the same kwargs as `numpy.loadtxt()`, but in this case slightly different behaviour is probably needed: i.e. "lead characters / strings that will result in *not* reading the line" v.s. "lead characters / strings that *will* result in reading the line", and if the second is empty, then any line not prefixed with an entry in the first will be read. That would permit both reading the output of `transformcalc decompose` (which it seems to do currently), and reading key-values from commented headers (which I suspect was my cognitively dissonant intention).